src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp

```diff
@@ -835,8 +828,6 @@ bool CachedOnDiskReadBufferFromFile::nextImplStep()
 
     chassert(!internal_buffer.empty());
 
-    swap(*implementation_buffer);
```

Comment:
Ah, I see, `use_external_buffer` in ReadBufferFromS3 is enabled, and it means that we have to assign its `internal_buffer`.

Would it still be better to `implementation_buffer->set(...)` (like `ReadBufferFromS3` does with its `internal_buffer`) instead of `swap()`, to worry less about missing a matching `swap()` later (on exception or early return)?

The other `swap()` wasn't in SCOPE_EXIT; wouldn't that leave the CachedOnDiskReadBufferFromFile in invalid state, with internal_buffer empty? I guess that's fine because we don't expect the CachedOnDiskReadBufferFromFile to be used after its `nextImpl()` throws?

Is this a general convention for all ReadBuffers - that they shouldn't be used after nextImple() threw? This doesn't seem to hold for `FileSegment::remote_file_reader` - it can be reused by a later CachedOnDiskReadBufferFromFile even if the ReadBuffer threw an exception into an earlier CachedOnDiskReadBufferFromFile; right? Would it make sense to clarify the contract for what can be done to ReadBuffers after they threw, and add a comment in ReadBuffer? Or it it deliberately underspecified?

Reply:
> to worry less about missing a matching swap() later (on exception or early return)

btw there is a `swapHelper` struct somewhere in the code for this purpose

> wouldn't that leave the CachedOnDiskReadBufferFromFile in invalid state, with internal_buffer empty?

I do not think so, it will mean that we will lose the buffer which we passed through to be later passed to impl, but if we get an exception - we will not need this buffer anymore

>  I guess that's fine because we don't expect the CachedOnDiskReadBufferFromFile to be used after its nextImpl() throws?

yes

> Would it still be better to implementation_buffer->set(...)

so it is also an option, but not sure if it will make the difference

> This doesn't seem to hold for FileSegment::remote_file_reader - it can be reused by a later CachedOnDiskReadBufferFromFile even if the ReadBuffer threw an exception into an earlier CachedOnDiskReadBufferFromFile; right?

hm, yes, but as I understand it can result in a problem only in case we successfully read via this reader and then failed to write the result into file segment, but this write fails (or if the command before write - reserve - fails) then file_segment's state is changed to the one, which will not allow any further usage of FileSegment::remote_reader. There is no operation in between these events, so I assume this case will be processed correctly.